lib/decompress.ps1

```diff
@@ -55,206 +56,275 @@ function Expand-7zipArchive {
-        try {
-            $7zPath = (Get-Command '7z' -CommandType Application | Select-Object -First 1).Source
-        } catch [System.Management.Automation.CommandNotFoundException] {
-            abort "Cannot find external 7-Zip (7z.exe) while '7ZIPEXTRACT_USE_EXTERNAL' is 'true'!`nRun 'scoop config 7ZIPEXTRACT_USE_EXTERNAL false' or install 7-Zip manually and try again."
-        }
-    } else {
-        $7zPath = Get-HelperPath -Helper 7zip
-    }
-    $LogPath = "$(Split-Path $Path)\7zip.log"
-    $ArgList = @('x', "`"$Path`"", "-o`"$DestinationPath`"", '-y')
-    $IsTar = ((strip_ext $Path) -match '\.tar$') -or ($Path -match '\.t[abgpx]z2?$')
-    if (!$IsTar -and $ExtractDir) {
-        $ArgList += "-ir!$ExtractDir\*"
-    }
-    if ($Switches) {
-        $ArgList += (-split $Switches)
-    }
-    switch ($Overwrite) {
-        "All" { $ArgList += "-aoa" }
-        "Skip" { $ArgList += "-aos" }
-        "Rename" { $ArgList += "-aou" }
-    }
-    $Status = Invoke-ExternalCommand $7zPath $ArgList -LogPath $LogPath
-    if (!$Status) {
-        abort "Failed to extract files from $Path.`nLog file:`n  $(friendly_path $LogPath)`n$(new_issue_msg $app $bucket 'decompress error')"
-    }
-    if (!$IsTar -and $ExtractDir) {
-        movedir "$DestinationPath\$ExtractDir" $DestinationPath | Out-Null
-    }
-    if (Test-Path $LogPath) {
-        Remove-Item $LogPath -Force
-    }
-    if ($IsTar) {
-        # Check for tar
-        $Status = Invoke-ExternalCommand $7zPath @('l', "`"$Path`"") -LogPath $LogPath
-        if ($Status) {
-            $TarFile = (Get-Content -Path $LogPath)[-4] -replace '.{53}(.*)', '$1' # get inner tar file name
-            Expand-7zipArchive -Path "$DestinationPath\$TarFile" -DestinationPath $DestinationPath -ExtractDir $ExtractDir -Removal
+    begin {
+        if ((get_config 7ZIPEXTRACT_USE_EXTERNAL)) {
+            try {
+                $7zPath = (Get-Command '7z' -CommandType Application | Select-Object -First 1).Source
+            } catch [System.Management.Automation.CommandNotFoundException] {
+                abort "Cannot find external 7-Zip (7z.exe) while '7ZIPEXTRACT_USE_EXTERNAL' is 'true'!`nRun 'scoop config 7ZIPEXTRACT_USE_EXTERNAL false' or install 7-Zip manually and try again."
+            }
         } else {
-            abort "Failed to list files in $Path.`nNot a 7-Zip supported archive file."
+            $7zPath = Get-HelperPath -Helper 7zip
         }
     }
-    if ($Removal) {
-        # Remove original archive file
-        Remove-Item $Path -Force
+    process {
+        if (!$DestinationPath) {
+            $DestinationPath = Split-Path -Path $Path
+        }
+        for ($i = 0; $i -lt $Path.Length; $i++) {
+            $aPath = $Path[$i]
+            if ($aDestinationPath = $DestinationPath[$i]) {
+                $aDestinationPath = $DestinationPath[-1]
+            }
+            if ($ExtractDir) {
+                if ($aExtractDir = $ExtractDir[$i]) {
```

Comment:
What are these check for?? It does not make sense at all.

Why do you need to check if assigning to variable was possible? For index out of bounds??

Sidenote: Naming `$aArray` is completely misleading and non-intuitive. Rather go with `currentDestination`

Reply:
That's for index out of bounds, e.g., `Expand-xxx -Path @(xxx) -Destination xxx` will extract all archives into one location, so it use `$DestinationPath[-1]`.

Assigning variable will be done without error, but `$aDestinationPath` may be `$null`, the `if` is checking not null. i.e., it is shorten for

```json
$aDestinationPath = $DestinationPath[$i]
if ($null -eq $aDestinationPath) {
    $aDestinationPath = $DestinationPath[-1]
}
```

I use `$aXXXXX` for noting singular variable, 'a DestinationPath'. Maybe `current` is better except for too long variable name...

Reply:
What is wrong with?

```powershell
if ($null -eq $DestinationPath[$i]) {
    .....
}
```

Reply:
That's okay, too. You mean?

```json
if ($null -eq $DestinationPath[$i]) {
    $aDestinationPath = $DestinationPath[$i]
} else {
    $aDestinationPath = $DestinationPath[-1]
}
```